src/app/qgshandlebadlayers.cpp

```diff
@@ -364,79 +371,48 @@ void QgsHandleBadLayers::editAuthCfg()
 void QgsHandleBadLayers::apply()
 {
   QgsProject::instance()->layerTreeRegistryBridge()->setEnabled( true );
-  QHash<QString, QString> baseChange;
-
 
   for ( int i = 0; i < mLayerList->rowCount(); i++ )
   {
     int idx = mLayerList->item( i, 0 )->data( Qt::UserRole ).toInt();
     QDomNode &node = const_cast<QDomNode &>( mLayers[ idx ] );
 
-    const QString name = mLayerList->item( i, 0 )->text();
     QTableWidgetItem *item = mLayerList->item( i, 4 );
     QString datasource = item->text();
-    const QString basepath = datasource.left( datasource.lastIndexOf( '/' ) );
-    bool changed = false;
-
-    if ( mFileBase[ name ].size() == 1 )
-    {
-      if ( mFileBase[ name ][0] != basepath && !baseChange.contains( mFileBase[ name ][0] ) )
-      {
-        baseChange[ mFileBase[ name ][0] ] = basepath;
-        changed = true;
-      }
-    }
-    else if ( mFileBase[ name ].size() > 1 )
-    {
-      if ( mFileBase[ name ].indexOf( basepath ) == -1 )
-      {
-        const QList<QString> fileBases = mFileBase[ name ];
-        for ( QString fileBase : fileBases )
-        {
-          if ( !baseChange.contains( fileBase ) )
-          {
-            baseChange[ fileBase ] = basepath;
-            changed = true;
-          }
-        }
-      }
-    }
-    if ( !changed && baseChange.contains( basepath ) )
-      datasource = datasource.replace( basepath, baseChange[basepath] );
+    QString fileName;
+    const QString name { mLayerList->item( i, 0 )->text() };
+    const QString layerId { node.namedItem( QStringLiteral( "id" ) ).toElement().text() };
+    const QString basepath = datasource.left( datasource.lastIndexOf( QDir::separator() ) );
+    const QString longName = datasource.mid( datasource.lastIndexOf( QDir::separator() ) );
+    if ( longName.lastIndexOf( '|' ) != -1 )
```

Comment:
Instead of this approach, it would be much safer to use QgsProviderRegistry::decodeUri and extract the path component from the results -- that automatically handles stripping any layerName parameters. (Not all providers use "|" to identify these)

Reply:
I have a question, would this work even on invalid files. My aim is simply to remove the extra arguments passed, in order to only search for the file by name and then append them back to the filepath.

 Now when I look at the  provideregistry::decideUri (at least in python) I need to make an instance of QgsProviderRegistry as the function is not static. This means that I must have a valid filepath to make the providerregistry, I'm also not even sure what is going to be outputed at this point.

Reply:
Yes -- it'll work on all layers, because it delegates to the provider itself to interpret the path exactly as it would if it was trying to load the layer.

You would use QgsProviderRegistry::instance()->decodeUri( layer_uri, layer_provider_name ), and then test if the returned map includes a "path" entry. If it doesn't, it's not file based (maybe db, maybe url, ...). If it does, then that's the path this dialog needs to work with.

Reply:
Great, thanks a ton!